src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IListOfTConverter.cs

```diff
@@ -12,9 +13,18 @@ internal sealed class IListOfTConverter<TCollection, TElement>
         : IEnumerableDefaultConverter<TCollection, TElement>
         where TCollection : IList<TElement>
     {
+        // Used to store elements when TCollection is ValueType
+        [MaybeNull]
+        private TCollection _listValues;
```

Comment:
This is not thread-safe as the converter instance may be used on multiple threads. Also I imagine the problem also exists for other collection types like `IList`, `ICollection<T>` etc, not just `IList<T>`. Tests need to be added for all these scenarios.

Reply:
@layomia I've just added tests following your recommendation.